pkg/webhook/server.go

```diff
@@ -369,34 +373,81 @@ func (wh *webhook) handleWebhookValidate(w http.ResponseWriter, req *http.Reques
 }
 
-func reviewResponse(uid types.UID, err error, aletmanagerHost string, resource string, name string, namespace string) *admissionv1.AdmissionReview {
-	// Currently hard coded and commented logic in order to just audit, but in order to get the content we need the policy to be deny.
-	// allowed := err == nil
-	//var status int32 = http.StatusAccepted
-	// if err != nil {
-	// 	status = http.StatusForbidden
-	// }
+func getValidationAnnotations(attrs admission.Attributes) (audit bool, deny bool) {
+	validationActionsPattern := `validationActions":\[(.*?)\]`
+	regex, _ := regexp.Compile(validationActionsPattern)
+
+	match := regex.FindStringSubmatch(fmt.Sprintf("%+v", attrs))
+	if len(match) >= 2 {
+		actions := match[1]
+		audit = strings.Contains(actions, "Audit")
+		deny = strings.Contains(actions, "Deny")
+	}
+
+	logger.Info("The actions are", "audit", audit, "deny", deny)
+
+	return audit, deny
+}
+
+func getMessage(attrs admission.Attributes) (message string) {
+	validationMessagePattern := `message":"(.*?)"`
+	regex, _ := regexp.Compile(validationMessagePattern)
+
+	match := regex.FindStringSubmatch(fmt.Sprintf("%+v", attrs))
+	if len(match) >= 2 {
+		message = match[1]
+	}
+
+	logger.Info("The message is", "message", message)
+
+	return message
+}
+
+func getPolicy(attrs admission.Attributes) (policy string) {
+	policyPattern := `policy":"(.*?)"`
+	regex, _ := regexp.Compile(policyPattern)
+
+	match := regex.FindStringSubmatch(fmt.Sprintf("%+v", attrs))
+	if len(match) >= 2 {
+		policy = match[1]
+	}
+
+	logger.Info("The policy is", "policy", policy)
+
+	return policy
+}
+
+func reviewResponse(uid types.UID, err error, aletmanagerHost string, resource string, name string, namespace string, attrs admission.Attributes) *admissionv1.AdmissionReview {
+	allowed := err == nil
+	var status int32 = http.StatusAccepted
+	if err != nil {
+		status = http.StatusForbidden
+	}
 	reason := metav1.StatusReasonUnknown
 	message := "valid"
-	// if err != nil {
-	// 	message = err.Error()
-	// }
+	if err != nil {
+		message = err.Error()
+	}
 
 	var statusErr *k8serrors.StatusError
 	if ok := errors.As(err, &statusErr); ok {
 		reason = statusErr.ErrStatus.Reason
 		message = statusErr.ErrStatus.Message
-		//status = statusErr.ErrStatus.Code
+		status = statusErr.ErrStatus.Code
+	}
+
+	audit, deny := getValidationAnnotations(attrs)
+	if audit || deny {
 		if aletmanagerHost != "" {
-			policyName := regexp.MustCompile(`ValidatingAdmissionPolicy '([^']+)'`).FindStringSubmatch(message)
+			policyName := getPolicy(attrs)
 			alerter := alertmanager.New(aletmanagerHost, "")
 			alertInfo := alertmanager.AlertInfo{
-				Name:        fmt.Sprintf("Failed Policy: %v", policyName[1]),
+				Name:        fmt.Sprintf("Failed Policy: %v", policyName),
 				Severity:    string(reason),
 				Resource:    resource,
 				Instance:    name,
 				Namespace:   namespace,
-				Description: message,
+				Description: getMessage(attrs),
 			}
 			alerter.Alert(&alertInfo)
```

Comment:
I think you can have multiple annotations containing "audit" messages.

Think about the following scenario. I have two policies, both of them are bound (policy binding) with audit action to the same object that is being admitted.

In this case I think you will have two audit messages here.

If I am right, then we should capture both and send audit alert message for each

Reply:
That's a good point, I will test it and see what we get.